File: packages/dart/lib/src/reporter/markdown.dart
/// The report as a document: a heading per section and a table of findings.
library;

import 'package:chki18n/src/constants.dart';
import 'package:chki18n/src/reporter/context.dart';
import 'package:chki18n/src/reporter/text.dart';

/// A cell may hold a translated value, and a `|` in one would split the row.
String _cell(String text) => text.replaceAll('|', r'\|');

String _code(String text) => text.isEmpty ? '' : '`${_cell(text)}`';

/// A table with its columns padded in the source, which is how every Markdown
/// file in this project is written and what keeps the raw text readable when
/// whatever renders it does not.
List<String> _table(List<String> header, List<List<String>> rows) {
  final widths = [
    for (var column = 0; column < header.length; column += 1)
      widestOf([header[column], for (final row in rows) row[column]]),
  ];

  String line(List<String> cells) =>
      '| ${[for (var column = 0; column < cells.length; column += 1) padTo(cells[column], widths[column])].join(' | ')} |';

  return [
    line(header),
    '| ${[for (final width in widths) '-' * width].join(' | ')} |',
    for (final row in rows) line(row),
  ];
}

/// The report as a document, for a pull request comment or a report checked in
/// beside the translations.
String formatMarkdown(Chki18nReportContext context) {
  final result = context.result;
  final lines = <String>[
    '# Translation check',
    '',
    '**${countsSentence(result.summary)}** ${scopeSentence(result)} '
        'Compared against `${result.target}`.',
  ];

  final byCode = context.options.groupBy == Chki18nGroupBy.code;

  for (final section in context.sections) {
    lines.addAll(['', '## ${section.label.isEmpty ? 'Issues' : section.label}', '']);
    lines.addAll(
      _table(
        ['Level', byCode ? 'Locale' : 'Check', 'Key', 'Value', 'Note'],
        [
          for (final issue in section.issues)
            [
              issue.level.name,
              _code(byCode ? issue.locale : issue.code.code),
              _code(keyLabelOf(issue, context.showGroup)),
              _cell(referenceOf(issue, result.target)),
              _cell(detailOf(issue)),
            ],
        ],
      ),
    );
  }

  if (context.sections.isEmpty) {
    lines.addAll(['', 'Nothing to report.']);
  }

  if (context.hidden > 0) {
    lines.addAll([
      '',
      '${plural(context.hidden, 'issue')} not shown, because of the level options.',
    ]);
  }

  return lines.join('\n');
}

File: packages/javascript/lib/reporter/markdown.ts
import { GROUP_BY } from '../constants.js';
import {
	countsSentence,
	detailOf,
	keyLabelOf,
	padTo,
	plural,
	referenceOf,
	scopeSentence,
	widestOf
} from './text.js';
import type { Chki18nReportContext } from './context.js';

/** A cell may hold a translated value, and a `|` in one would split the row. */
const cell = (text: string): string => text.replace(/\|/g, '\\|');

const code = (text: string): string => (text ? `\`${cell(text)}\`` : '');

/**
 * A table with its columns padded in the source, which is how every Markdown
 * file in this project is written and what keeps the raw text readable when
 * whatever renders it does not.
 */
function table(header: string[], rows: string[][]): string[] {
	const widths = header.map((title, column) =>
		widestOf([title, ...rows.map((row) => row[column])])
	);
	const line = (cells: string[]) =>
		`| ${cells.map((text, column) => padTo(text, widths[column])).join(' | ')} |`;

	return [
		line(header),
		`| ${widths.map((width) => '-'.repeat(width)).join(' | ')} |`,
		...rows.map((row) => line(row))
	];
}

/**
 * The report as a document: a heading per section and a table of findings, for
 * a pull request comment or a report checked in beside the translations.
 */
export function formatMarkdown(context: Chki18nReportContext): string {
	const { options, result, sections } = context;
	const lines: string[] = [
		'# Translation check',
		'',
		`**${countsSentence(result.summary)}** ${scopeSentence(result)} Compared against \`${result.target}\`.`
	];

	const byCode = options.groupBy === GROUP_BY.CODE;

	for (const section of sections) {
		lines.push('', `## ${section.label || 'Issues'}`, '');
		lines.push(
			...table(
				['Level', byCode ? 'Locale' : 'Check', 'Key', 'Value', 'Note'],
				section.issues.map((issue) => [
					issue.level,
					code(byCode ? issue.locale : issue.code),
					code(keyLabelOf(issue, context.showGroup)),
					cell(referenceOf(issue, result.target)),
					cell(detailOf(issue))
				])
			)
		);
	}

	if (sections.length < 1) {
		lines.push('', 'Nothing to report.');
	}

	if (context.hidden > 0) {
		lines.push('', `${plural(context.hidden, 'issue')} not shown, because of the level options.`);
	}

	return lines.join('\n');
}
